airflow/opt/providers/etna/etna/hooks/c4.py

```diff
@@ -95,48 +98,47 @@ def __init__(self, hook: C4Hook):
-        file_size: int
+        source_file: SftpEntry,
+        cat_hook: CatHook
      ):
         """
         Stream the file to C4, to the given path.
 
         params:
+          sftp: a connected SFTPClient instance
           folder_path: str, the destination path on C4 (relative to the hook's configured root_path)
           file_name: str, the file name to save to, on C4
-          file_obj: io, the file-like object that is a data source
-          file_size: int, size of the file in bytes
+          source_file: SftpEntry, the source file to download
+          cat_hook: CatHook, the CAT hook with SFTP login information
         """
-        def _log_progress(consumed: int, total: int):
-            # Only log every 20 seconds, to save log space...
-            nonlocal should_log
-            time_check = int(time.time())
-            if time_check % 20 == 0 and should_log:
-                log.info("Upload progress: " + "{:.2%}".format(consumed / total))
-                should_log = False
-            elif time_check % 20 != 0 and not should_log:
-                should_log = True
-
-        with self.sftp() as sftp:
-            # Ensure the folder path exists on C4
-            full_path = os.path.join(self._root_path(), folder_path)
-            traversed = []
-            for folder in full_path.split("/"):
-                if folder == "":
-                    continue
-
-                path_to_folder = os.path.join("/", "/".join(traversed), folder)
-                try:
-                    sftp.chdir(path_to_folder)
-                except IOError:
-                    sftp.mkdir(path_to_folder)
-                traversed.append(folder)
-
-            should_log = True
-            sftp.putfo(
-                file_obj,
-                os.path.join(full_path, file_name),
-                file_size=file_size,
-                callback=_log_progress)
\ No newline at end of file
+        # Ensure the folder path exists on C4
+        full_path = os.path.join(self._root_path(), folder_path)
+        traversed = []
+        for folder in full_path.split("/"):
+            if folder == "":
+                continue
+
+            path_to_folder = os.path.join("/", "/".join(traversed), folder)
+            try:
+                sftp.chdir(path_to_folder)
+            except IOError:
+                sftp.mkdir(path_to_folder)
+            traversed.append(folder)
+
+        cmd = f"lftp sftp://{cat_hook.connection.login}:{cat_hook.connection.password}@{cat_hook.connection.host}  -e \"get {source_file.full_path} -o {os.path.join(full_path, file_name)}; bye\""
+
+        with self.ssh() as ssh:
+            stdin, stdout, stderr = ssh.exec_command(cmd)
+
+            # exec_command is non-blocking, so we'll
+            #   add a blocking call here.
+            # We shouldn't deadlock since the LFTP output is small
+            # https://stackoverflow.com/a/28486051
+            exit_status = stdout.channel.recv_exit_status()
```

Comment:
I'd consider reading the solution here https://stackoverflow.com/a/31650750 just in case, it should be a small change.

So what's happening is that, when stdout is attached to a buffered output, it actually blocks when that buffer is full, under the assumption that some underlying reader needs to catch up.  stdout is designed to be loss-less; if there is more output than the consumer could keep up with, you only have two options: drop the extra output or block.  That's a principle of streaming.

This 'works' whenever the default buffer is enough to receive data from the stdout, but in general we should consume all the stdout if possible, even if we're throwing it away (I wouldn't actually print it, for instance, just read it).